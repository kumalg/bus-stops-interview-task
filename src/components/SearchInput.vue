<template>
    <div :class="classes">
        <label class="labelled-container">
            <span class="placeholder">{{ placeholder }}</span>
            <input
                class="input"
                type="text"
                :value="internalValue"
                @input="handleInput($event)"
                @focus="setFocus(true)"
                @blur="setFocus(false)"
            />
            <SearchIcon class="search-icon" />
        </label>
    </div>
</template>

<script setup lang="ts">
import { computed, ref } from 'vue';

import { debounce } from '@/assets/debounce';

import SearchIcon from '@/icons/SearchIcon.vue';

const emit = defineEmits<{
    (e: 'update:model-value', value: string): void;
}>();

const props = withDefaults(
    defineProps<{
        modelValue: string;
        placeholder?: string;
    }>(),
    {
        placeholder: 'Search'
    }
);

const focused = ref(false);
const internalValue = ref(props.modelValue);

const classes = computed(() => [
    'search-input',
    { focused: focused.value, filled: !!internalValue.value }
]);

const emitModelUpdate = debounce((value: string) => {
    emit('update:model-value', value);
}, 250);

const handleInput = (event: Event) => {
    const el = event.target as HTMLInputElement;
    internalValue.value = el.value;
    emitModelUpdate(el.value);
};

const setFocus = (value: boolean) => {
    focused.value = value;
};
</script>

<style lang="scss" scoped>
.search-input {
    .labelled-container {
        position: relative;
        display: block;
    }

    .placeholder,
    .search-icon {
        position: absolute;
        top: 50%;
        transform: translateY(-50%);
        user-select: none;
        pointer-events: none;
    }

    .placeholder {
        left: 0.75rem;
        color: $color-gray-700;
        background-color: $color-gray-000;
        padding: 0 0.25rem;

        @include default-transition('font-size, top');
    }

    .search-icon {
        right: 1rem;
        color: $color-gray-600;
        visibility: visible;

        @include default-transition('visibility, opacity');
    }

    .input {
        border: 1px solid $color-gray-500;
        border-radius: 0.25rem;
        height: 2.5rem;
        padding: 0 2.5rem 0 1rem;
        display: inline-flex;
        font-family: inherit;
        color: $color-gray-900;
        background-color: $color-gray-000;
        width: 100%;

        @include default-transition('border-color');

        &:hover {
            outline: none;
            border-color: $color-gray-600;
        }

        &:focus {
            outline: none;
            border-color: $color-accent;
        }
    }

    .placeholder,
    .input {
        font-size: 0.75rem;
        line-height: 1rem;
    }

    &.focused,
    &.filled {
        .placeholder {
            top: 0;
            font-size: 0.625rem;
        }

        .search-icon {
            opacity: 0;
            visibility: hidden;
        }

        .input {
            padding-right: 1rem;
        }
    }
}
</style>
