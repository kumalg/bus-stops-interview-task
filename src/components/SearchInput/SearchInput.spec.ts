import { flushPromises, shallowMount } from '@vue/test-utils';

import SearchInput from './SearchInput.vue';

describe('SearchInput', () => {
    const SearchInputFactory = (
        props: Partial<InstanceType<typeof SearchInput>['$props']> = {}
    ) => {
        return shallowMount(SearchInput, {
            props: {
                modelValue: '',
                ...props
            }
        });
    };

    it('renders proper html markup', () => {
        const wrapper = SearchInputFactory();

        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders native input element', () => {
        const wrapper = SearchInputFactory();

        expect(wrapper.find('input').exists()).toBe(true);
    });

    it('update model on input event inside native input element', async () => {
        const TEST_INPUT = 'test_input';
        const wrapper = SearchInputFactory({ debounce: 0 });
        const inputEl = wrapper.find('input');

        inputEl.setValue(TEST_INPUT);

        await flushPromises();

        expect(wrapper.emitted('update:model-value')?.[0]).toEqual([TEST_INPUT]);
    });

    it('not update model on input event with debounce before debounce defined time', async () => {
        const TEST_INPUT = 'test_input';
        const DEBOUNCE = 250;

        const wrapper = SearchInputFactory({ debounce: DEBOUNCE });
        const inputEl = wrapper.find('input');

        inputEl.setValue(TEST_INPUT);

        await flushPromises();

        expect(wrapper.emitted('update:model-value')).not.toBeDefined();
    });

    it('update model on input event with debounce after debounce defined time', (done) => {
        const TEST_INPUT = 'test_input';
        const DEBOUNCE = 250;

        const wrapper = SearchInputFactory({ debounce: DEBOUNCE });
        const inputEl = wrapper.find('input');

        inputEl.setValue(TEST_INPUT);

        expect(wrapper.emitted('update:model-value')).not.toBeDefined();

        setTimeout(() => {
            expect(wrapper.emitted('update:model-value')?.[0]).toEqual([TEST_INPUT]);
            done();
        }, DEBOUNCE);
    });

    describe('has additional classes', () => {
        it('focused class on focus', async () => {
            const wrapper = SearchInputFactory();
            const inputEl = wrapper.find('input');

            expect(wrapper.classes()).not.toContain('focused');

            await inputEl.trigger('focus');
            expect(wrapper.classes()).toContain('focused');

            await inputEl.trigger('blur');
            expect(wrapper.classes()).not.toContain('focused');
        });

        it('filled class when input has value', async () => {
            const wrapper = SearchInputFactory();
            const inputEl = wrapper.find('input');

            expect(wrapper.classes()).not.toContain('filled');

            await inputEl.setValue('example');
            expect(wrapper.classes()).toContain('filled');

            await inputEl.setValue('');
            expect(wrapper.classes()).not.toContain('filled');
        });
    });

    describe('placeholder', () => {
        it('render default when none provided in prop', () => {
            const wrapper = SearchInputFactory();

            expect(wrapper.find('.placeholder').text()).toBe('Search');
        });

        it('render custom when provided in prop', () => {
            const EXAMPLE_PLACEHOLDER = 'example-placeholder';
            const wrapper = SearchInputFactory({
                modelValue: '',
                placeholder: EXAMPLE_PLACEHOLDER
            });

            expect(wrapper.find('.placeholder').text()).toBe(EXAMPLE_PLACEHOLDER);
        });
    });
});
