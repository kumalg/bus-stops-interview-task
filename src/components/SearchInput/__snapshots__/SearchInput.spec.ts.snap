// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput renders proper html markup 1`] = `
<div
  class="search-input"
>
  <label
    class="labelled-container"
  >
    <span
      class="placeholder"
    >
      Search
    </span>
    <input
      class="input"
      type="text"
    />
    <search-icon-stub
      class="search-icon"
    />
  </label>
</div>
`;
